### ISSUE TITLE
ValueTypes: Withfeld instruction disallows access to private final in nest member

### ISSUE THREAD

**Issue by tajila** [eclipse/openj9#11084]
The resolution with withfield uses `J9_RESOLVE_FLAG_WITH_FIELD` as well as `J9_RESOLVE_FLAG_FIELD_SETTER`. The field setter flags follow the final field rules described  <a href="https://docs.oracle.com/javase/specs/jvms/se15/html/jvms-6.html#jvms-6.5.putfield">here </a>. 

`Otherwise, if the resolved field is final, it must be declared in the current class, and the instruction must occur in an instance initialization method of the current class. Otherwise, an IllegalAccessError is thrown.`
 
Withfield however must follow these rules:

`Otherwise, if the resolved field is not declared by a class that belongs to the same nest as the current class, any of the exceptions pertaining to the nestmate test (5.4.4) may be thrown.`

We need to modify the withfield <a href="https://github.com/eclipse/openj9/blob/master/runtime/vm/BytecodeInterpreter.hpp#L8662">bytecode</a> to pass the withfield resolve flag and not the field setter flag. The resolve code `resolveInstanceFieldRef` needs to be updated to perform nest check.

**Comment by tajila** [comment]
FYI @hzongaro

**Comment by tajila** [comment]
@hzongaro can you please point to where your test case is

@hangshao0 can you please take a look at this issue

**Comment by hangshao0** [comment]
> @hangshao0 can you please take a look at this issue

Yes, I will look at this.

**Comment by hzongaro** [comment]
> @hzongaro can you please point to where your test case is

@hangshao0 Please see pull request #11097 for the test:  `testUnresolvedWithFieldUse`

**Comment by hzongaro** [comment]
@hangshao0 I made a mistake which resulted in circular `NestHost` attributes being added to a classes.  I corrected this in commit fbba19114, which is the latest commit associated with pull request #11097

**Comment by hangshao0** [comment]
I have reproduced the IAE:

```
java.lang.IllegalAccessError: Class UsingUnresolvedD illegally accessing "private" member of class UnresolvedD
        at UsingUnresolvedD.testUnresolvedValueTypeWithField(UsingUnresolvedD.java)
        at org.openj9.test.lworld.ValueTypeTests.testUnresolvedWithFieldUse(ValueTypeTests.java:2803)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
        ...
```

However after I added the nestmate check, I still got the same IAE. 
With some printfs in the code, I see the nestHost of class `UsingUnresolvedD` is `UsingUnresolvedD`. The nest host of class `UnresolvedD` is `UnresolvedD`. So classes `UnresolvedD` and `UsingUnresolvedD` are in 2 different nests, the IAE seems to be the correct behaviour in this case. Is there something wrong with the test code ? @hzongaro

**Comment by hzongaro** [comment]
> The nest host of class `UnresolvedD` is `UnresolvedD`. So classes `UnresolvedD` and `UsingUnresolvedD` are in 2 different nests, the IAE seems to be the correct behaviour in this case. Is there something wrong with the test code ?

Hang @hangshao0, I think the problem with the circular nest host for `UsingUnresolvedD` was fixed in commit fbba191.  However, I see that there is still another problem - now `UnresolvedD` refers to a nest host that doesn't exist.  Sorry about that!

May I ask you to try out the changes to the test in draft pull request #11316?

**Comment by hangshao0** [comment]
> May I ask you to try out the changes to the test in draft pull request #11316?

I copied your change of ValueTypeTests.java from #11316, I see `testUnresolvedWithFieldUse` passed with my local VM change.
